notebooks/hfdemo/tspulse/anomaly_detection/utility/datasets.py

```diff
@@ -0,0 +1,242 @@
+        self.data = self._normalize_data(data) if normalize else data
+
+        if self.data.shape[0] < window_size:
+            pad_len = window_size - self.data.shape[0]
+            pad = np.ones((pad_len, self.data.shape[1])) * self.data[:1]
+            self.data = np.concatenate((pad, self.data), axis=0)
+
+        self.label = label
+        self.return_dict = return_dict
+        self.channel_last = channel_last
+
+        self.univariate = self.data.shape[1] == 1
+        self.sample_num = max(((self.data.shape[0] - window_size) // stride) + 1, 0)
+
+        if self.label is not None:
+            self.samples, self.gen_labels = self._generate_samples()
+        else:
+            self.samples = self._generate_samples()
+        self.input_mask = np.ones((self.window_size, data.shape[1]), dtype=np.float32)  # Fixed input mask
+        if aggr_window_size is not None:
+            self.input_mask[:aggr_window_size, :] = 0
+
+        if not self.channel_last:
+            self.samples = self.samples.permute(0, 2, 1)  # batch, channel, window_size
+            self.input_mask = self.input_mask[:, 0]  # window_size
+
+    def _normalize_data(self, data, epsilon=1e-8):
+        mean, std = np.mean(data, axis=0), np.std(data, axis=0)
+        std = np.where(std == 0, epsilon, std)  # Avoid division by zero
+        return (data - mean) / std
+
+    def _generate_samples(self):
+        data = torch.tensor(self.data, dtype=torch.float32)
+        indices = np.arange(0, self.sample_num * self.stride, self.stride)
+
+        if self.univariate:
+            X = torch.stack([data[i : i + self.window_size] for i in indices])
+            print(X.shape)
+        else:
+            X = torch.stack([data[i : i + self.window_size] for i in indices])
+            print(X.shape)
+
+        if self.label is not None:
+            self.label = torch.tensor(self.label)
+            Y = torch.stack([self.label[i : i + self.window_size] for i in indices])
+            return X, Y
+
+        return X
+
+    def __len__(self):
+        return self.sample_num
+
+    def __getitem__(self, index):
+        if self.label is not None:
+            if self.return_dict:
+                return {
+                    "past_values": self.samples[index],
+                    "anomaly_labels": self.gen_labels[index],
+                    "past_observed_mask": self.input_mask,
+                }
+            else:
+                return self.samples[index], self.input_mask, self.gen_labels[index]
+        else:
+            if self.return_dict:
+                return {
+                    "past_values": self.samples[index],
+                    "past_observed_mask": self.input_mask,
+                }
+            else:
+                return self.samples[index], self.input_mask
+
+
+# The following dataset is used by TSPulse
+# It generates data for reconstruction as well as
+# forecasting task.
+class TSPulseForecastDataset(torch.utils.data.Dataset):
```

Comment:
Similar comment as above -- not sure why we can't use existing approaches. There is some duplicated code here.